working/augmentation-libraries/parts_with_imports.md

```diff
@@ -0,0 +1,631 @@
+
+    *   Because of that, it’s possible to convert an existing library into a
+        part file of another library. Since a library is self-contained and
+        imports all external names that it refers to, making it a part file will
+        not cause any conflict due to inherited imports. _(Obviously still need
+        to avoid conflicts with top-level declarations.)_
+    *   And similarly, if a part file *is* self-contained, it can be converted
+        into a separate library and imported back into the original library, or
+        it can be moved to another position in the part tree hierarchy. _(Again
+        augmentations introduce complications, which is why it’s usually a good
+        idea to keep all augmentations inside the same part sub-tree).
+
+### Grammar
+
+We extend the grammar of part files to allow `import`, `export` and `part` file
+directives. We allow `part` files directives to use a configurable URI like the
+other two. We restrict the `part of` directive to only allow the string version.
+
+```ebnf
+-- Changed "<uri>" to "<configurableUri>".
+<partDirective> ::= <metadata> `part' <configurableUri> `;'
+
+-- Removed "<dottedIdentifier>" as option, retaining only "<uri>".
+<partHeader> ::= <metadata> `part' `of' <uri> `;'
+
+-- Added "<importOrExport>* <partDirective>*"
+<partDeclaration> ::=
+  <partHeader> <importOrExport>* <partDirective>* (<metadata>
+  <topLevelDeclaration>)* <EOF>
+```
+
+The grammar change is small, mainly adding `import`, `export` and `part`
+directives to part files.
+
+The change to `part of` directives to not allow a dotted name was made because
+we want a part file of a part file to refer back to its parent part file, but a
+dotted library name can only refer to a library. _That doesn’t mean that
+part-of-dotted-name cannot be supported for part files that are part files of a
+library file. It’s also that the Dart team wants to remove the feature, and has
+been linting against its use for quite a while already. Dotted names in part-of
+being partially incompatible with the new feature just means that now is a good
+opportunity to get rid of them._
+
+The change to a configurable URI for `part` files was made because it can ease
+one of the shortcomings of using libraries for platform-dependent code: That
+other libraries cannot provide implementations for private members, or code
+that accesses private members, without duplicating the entire library. With
+part files having their own imports, adding configurable URIs for `part`
+directives gives a way to avoid that code duplication, possibly even more
+conveniently if also using augmentations.
+
+The configurable URI for a `part` works just as for imports and exports, it
+chooses the URI that the `part` directive refers to, and after that the
+included file works just as any other part file.
+
+It’s a **compile-time error** if a Dart (parent) file with URI *P* has a `part`
```

Comment:
When I read this paragraph, it seems that there are a few typos.

"does not parse as a `<partDirective>`" - maybe `< partDeclaration >`?

"does not denote the library of *P*" - we can have `partDirective`s not only in libraries, so maybe "the file with URI *P*"?

Reply:
ACK, `<partDeclaration>` for the entire part file, `<partDirective>` for the `part "...";` directive.
And yes, "the file *P*" is correct. (Damn! I hoped I had found all of the places that assumed a parent file was a library.)